stdlib/public/core/StringGuts.swift

```diff
@@ -18,266 +18,382 @@ import SwiftShims
 // FIXME: what about ppc64 and s390x?
 //
 
+// TODO: Comments. Supposed to abstract bit-twiddling operations
 @_fixed_layout
-public // FIXME
-struct _StringGuts {
-  // TODO 32-bit: Insert padding between fields
-  public // FIXME for testing only
-  var _storage: (_BuiltinBridgeObject, UInt)
+public // @testable
+struct _StringObject {
+  // TODO: Proper built-in string object support. For now, we use BridgeObject
+  // which might be very slightly suboptimal and different than our bit
+  // patterns, but provides us the runtime functionality we want.
+  var _object: _BuiltinBridgeObject
 
   @_inlineable
-  public
-  var _object: _BuiltinBridgeObject {
-    @inline(__always) get { return _storage.0 }
-    @inline(__always) set { _storage.0 = newValue }
+  @inline(__always)
+  internal
+  init(_ object: _BuiltinBridgeObject) {
+    self._object = object
   }
+
   @_inlineable
-  public
-  var _otherBits: UInt {
-    @inline(__always) get { return _storage.1 }
-    @inline(__always) set { _storage.1 = newValue }
+  internal
+  var rawBits: UInt {
+    @inline(__always)
+    get { return _bitPattern(_object) }
   }
+}
+
+//
+// Constant bits and masks
+//
+//
+// Bit layout (x86-64 and arm64):
+//
+// _StringObject:
+// +---+---+---|---+------+----------------------------------------------------+
+// + t | v | o | w | uuuu | payload (54 bits)                                  |
+// +---+---+---|---+------+----------------------------------------------------+
+// where t: is-a-value, i.e. a tag bit that says not to perform ARC
+//       v: sub-variant bit, i.e. set for isCocoa or isSmall
+//       o: is-opaque, i.e. opaque vs contiguously stored strings
+//       w: width indicator bit (0: ASCII, 1: UTF-16)
+//       u: unused bits
+//
```

Comment:
And yes, I considered using `s: size indicator` instead of `w` so that it would spell `t|v|o|s`.